### PR TITLE
Trie nodes chunk processor

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added the trie nodes chunk processor. This processor is able to hold the completeness of a large trie nodes set and also is able to trigger requests for the missing parts.

### CODE REVIEW

File: process/interceptors/processor/trieNodeChunksProcessor.go [hunk 1]
```diff
@@ -1,3 +1,214 @@
...
+	}
+
+	respChan := make(chan process.CheckedChunkResult, 1)
+	req := checkRequest{
+		batch:        b,
+		chanResponse: respChan,
+	}
+
+	proc.chanCheckRequests <- req
+	response := <-respChan
```
Comment (iulianpascalau) [inline process/interceptors/processor/trieNodeChunksProcessor.go:121]: add a select here as to not wait indefinitely if the context is done and the respChan will never hold a value.

File: process/interceptors/processor/trieNodeChunksProcessor.go [hunk 2]
```diff
@@ -1,3 +1,214 @@
...
+
+	buff := chunkData.TryAssembleAllChunks()
+	haveAllChunks := len(buff) > 0
+	if haveAllChunks {
+		proc.chunksCacher.Remove(cr.batch.Reference)
+	} else {
+		proc.chunksCacher.Put(cr.batch.Reference, chunkData, chunkData.Size())
+	}
+
+	cr.chanResponse <- process.CheckedChunkResult{
```
Comment (iulianpascalau) [inline process/interceptors/processor/trieNodeChunksProcessor.go:146]: select with empty defer

File: process/interceptors/processor/chunk/chunk.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	if !gotAllParts {
+		//TODO(iulian) remove this log
+		log.Warn("not all parts got", "max chunks", c.maxChunks, "len chunk", len(c.data))
+		return nil
+	}
+
+	buff := make([]byte, 0, c.size)
+	for i := uint32(0); i < c.maxChunks; i++ {
+		part := c.data[i]
+
```
Comment (sasurobert) [inline process/interceptors/processor/chunk/chunk.go:48]: delete empty line

File: process/interceptors/multiDataInterceptor.go [hunk 4]
```diff
@@ -127,7 +127,7 @@ func (mdi *MultiDataInterceptor) ProcessReceivedMessage(message p2p.MessageP2P,
 	mdi.mutChunksProcessor.RUnlock()
 	if err != nil {
 		mdi.throttler.EndProcessing()
-		return nil
+		return err
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/interceptors/multiDataInterceptor.go:130]: No, this is the new code only on this feature branch. Will not affect consensus as this error is considered a programming error. Found out in unit tests that we should return error here.

File: process/interceptors/processor/chunk/chunk.go (hunk 3 above)
Comment (iulianpascalau) [reply process/interceptors/processor/chunk/chunk.go:48]: done

File: dataRetriever/requestHandlers/requestHandler.go [hunk 5]
```diff
@@ -443,20 +444,27 @@ func (rrh *resolverRequestHandler) RequestTrieNodes(destShardID uint32, hashes [
 	rrh.trieHashesAccumulator = make(map[string]struct{})
 }
 
+func (rrh *resolverRequestHandler) createIdentifier(requestHash []byte, chunkIndex uint32) []byte {
+	chunkBuffer := make([]byte, 4)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply dataRetriever/requestHandlers/requestHandler.go]: done

File: process/interceptors/processor/trieNodeChunksProcessor.go (hunk 2 above)
Comment (iulianpascalau) [reply process/interceptors/processor/trieNodeChunksProcessor.go:146]: done

File: process/interceptors/processor/trieNodeChunksProcessor.go (hunk 1 above)
Comment (iulianpascalau) [reply process/interceptors/processor/trieNodeChunksProcessor.go:121]: done, thanks

File: process/errors.go [hunk 6]
```diff
@@ -1009,3 +1009,9 @@ var ErrNilNumConnectedPeersProvider = errors.New("nil number of connected peers
 
 // ErrNilChunksProcessor signals that a nil chunks processor has been provided
 var ErrNilChunksProcessor = errors.New("nil chunks processor")
+
+// ErrIncompatibleReference signals that an incompatible reference
```
Comment (BeniaminDrasovean) [inline process/errors.go:1013]: Remove 'that'

File: process/interceptors/processor/chunk/chunk.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+// NewChunk creates a new chunk instance able to account for the existing and missing chunks of a larger buffer
+// Not a concurrent safe component
+func NewChunk(maxChunks uint32) *chunk {
+	return &chunk{
+		data:      make(map[uint32][]byte),
+		maxChunks: maxChunks,
+	}
+}
+
+// Put will add are rewrite an existing chunk
```
Comment (BeniaminDrasovean) [inline process/interceptors/processor/chunk/chunk.go:24]: `and` instead of `are`

File: process/errors.go (hunk 6 above)
Comment (iulianpascalau) [reply process/errors.go:1013]: rephrased

File: process/interceptors/processor/chunk/chunk.go (hunk 7 above)
Comment (iulianpascalau) [reply process/interceptors/processor/chunk/chunk.go:24]: changed to `or` as that was the original intent

File: process/interceptors/processor/trieNodeChunksProcessor.go [hunk 8]
```diff
@@ -1,3 +1,233 @@
...
+
+	return tncp, nil
+}
+
+func (proc *trieNodeChunksProcessor) processLoop(ctx context.Context) {
+	chanDoRequests := time.After(proc.requestInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("trieNodeChunksProcessor.processLoop go routine is stopping...")
```
Comment (sasurobert) [inline process/interceptors/processor/trieNodeChunksProcessor.go:94]: do you really want this debug ? :) if you want to test it real time, you should delete the protection code from smartContract process.go

**Review by sasurobert — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

File: process/interceptors/processor/trieNodeChunksProcessor.go (hunk 8 above)
Comment (iulianpascalau) [reply process/interceptors/processor/trieNodeChunksProcessor.go:94]: We have debug prints like this on all go routines as to easily check during debugging sessions

**Pull request merged**